File: test/tools/centre.js
const { launch, SITE } = require("../browser");
const fs=require('fs');
/* Does the poem sit centred between the screen's edge and her shirt?
   The shirt is found in the RENDERED pixels — bright and nearly
   neutral, which is the cotton and not the warm streaks across it —
   rather than trusted from the formula that positions the block. */
(async()=>{
  const b=await launch();
  const SIZES=[[1280,800],[1366,768],[1440,900],[1600,900],[1680,1050],[1920,1080],
               [2560,1440],[1280,1024],[1400,1050],[1300,760]];
  const out=[];
  for(const [w,h] of SIZES){
    const ctx=await b.newContext({viewport:{width:w,height:h}});
    const pg=await ctx.newPage();
    await pg.goto(SITE + '/index.html#about',{waitUntil:'load'});
    await pg.waitForTimeout(1800);
    const geo=await pg.evaluate(()=>{
      const ps=[...document.querySelectorAll('.hand p')].slice(0,4);
      const l=Math.min(...ps.map(p=>p.getBoundingClientRect().left));
      const r=Math.max(...ps.map(p=>p.getBoundingClientRect().right));
      const cs=getComputedStyle(document.querySelector('#about'));
      return {left:Math.round(l),right:Math.round(r),
              padLeft:cs.paddingLeft, shirtCalc:cs.getPropertyValue('--shirt')};
    });
    await pg.screenshot({path:`${process.env.SP}/c-${w}x${h}.png`});
    out.push([w,h,geo]);
    await ctx.close();
  }
  await b.close();
  fs.writeFileSync(process.env.SP+'/centre.json',JSON.stringify(out));
  console.log('measured');
})();

File: test/tools/centre.py
#!/usr/bin/env python3
"""PASS/FAIL for the poem's placement. Run scratchpad/centre.js first;
it writes the block geometry and one screenshot per size. The shirt is
found in the RENDERED pixels -- bright and nearly neutral, which is the
cotton and not the warm streaks across it -- so this checks the result
rather than re-deriving the formula that produced it."""
from PIL import Image
import numpy as np, json, sys, os
SP=os.path.dirname(os.path.abspath(__file__))
data=json.load(open(SP+"/centre.json"))
bad=0
print(f"{'size':>12}  {'gap left':>9} {'gap right':>10}  {'diff':>6}   shirt at")
for w,h,g in data:
    a=np.array(Image.open(f"{SP}/c-{w}x{h}.png").convert("RGB")).astype(int)
    lum=(a[...,0]*299+a[...,1]*587+a[...,2]*114)//1000
    sat=a.max(axis=2)-a.min(axis=2)
    cols=((lum>195)&(sat<38)).sum(axis=0)
    shirt=min(x for x in range(w) if cols[x]>12)
    gl=g["left"]; gr=shirt-g["right"]; d=gl-gr
    ok=abs(d)<=3
    if not ok: bad+=1
    print(f"{'PASS' if ok else 'FAIL'} {w}x{h:<7} {gl:9d} {gr:10d}  {d:+6d}   {shirt:6d}")
print(f"\n{bad} FAILED" if bad else "\nALL PASS")
sys.exit(1 if bad else 0)
